src/array/cpu/traversal.cc

```diff
@@ -0,0 +1,242 @@
+/*!
+ *  Copyright (c) 2020 by Contributors
+ * \file array/cpu/traversal.cc
+ * \brief Graph traversal implementation
+ */
+
+#include <dgl/graph_traversal.h>
+#include <algorithm>
+#include <queue>
+#include "./traversal.h"
+
+namespace dgl {
+namespace aten {
+namespace impl {
+namespace {
+// A utility view class to wrap a vector into a queue.
+template<typename DType>
+struct VectorQueueWrapper {
+  std::vector<DType>* vec;
+  size_t head = 0;
+
+  explicit VectorQueueWrapper(std::vector<DType>* vec): vec(vec) {}
+
+  void push(const DType& elem) {
+    vec->push_back(elem);
+  }
+
+  DType top() const {
+    return vec->operator[](head);
+  }
+
+  void pop() {
+    ++head;
```

Comment:
Will `head` somehow goes beyond `vec->size()`?

Reply:
Currently, it will not. It is used by the traversal funcs and will not goes beyoud vec->size() in these cases.